src/hotspot/share/runtime/os.cpp

```diff
@@ -2105,18 +2105,23 @@ void os::pretouch_memory(void* start, void* end, size_t page_size) {
     // We're doing concurrent-safe touch and memory state has page
     // granularity, so we can touch anywhere in a page.  Touch at the
     // beginning of each page to simplify iteration.
-    char* cur = static_cast<char*>(align_down(start, page_size));
-    void* last = align_down(static_cast<char*>(end) - 1, page_size);
-    assert(cur <= last, "invariant");
-    // Iterate from first page through last (inclusive), being careful to
+    void* aligned_start = align_down(start, page_size);
+    void* aligned_end = align_up(end, page_size);
+    assert(aligned_start < aligned_end, "invariant");
+    // Iterate from first page to the aligned end (exclusive), being careful to
     // avoid overflow if the last page abuts the end of the address range.
-    for ( ; true; cur += page_size) {
-      Atomic::add(reinterpret_cast<int*>(cur), 0, memory_order_relaxed);
-      if (cur >= last) break;
-    }
+    pd_pretouch_memory(aligned_start, aligned_end, page_size);
   }
 }
 
+void os::pretouch_memory_fallback(void *start, void *end, size_t page_size) {
+  char *cur = static_cast<char *>(start);
+  do {
+    Atomic::add(reinterpret_cast<int *>(cur), 0, memory_order_relaxed);
+    cur += page_size;
+  } while (cur < end);
```

Comment:
The previous code was careful to touch only as far as `align_down(static_cast<char*>(end) - 1, page_size)` to avoid the case where the region to be pre-touched ended at the end of the address space.  (There is a comment about that.)  Your code might compute `cur + page_size` and wrap around to the beginnng of the address space.  You do not want to debug that, or have someone curse you for making them debug it.

Your code might find that `cur < end` is never true if `end == 0` because you do not do the `-1` and the `align_down` and the comparison is unslgned.

A style comment: I usually write loops like the old line 2113 as `for ( ; /* break *; cur += page_size) {` to warn the reader to watch for a `break` from the loop.

Reply:
Thanks to point out the risk of overflow. I think it originates from the call to `align_up` earlier, and the following assertion would guarantee the computation of `cur + page_size` if overflow did not happen before. So I took back the original logic of inclusive ranges here, and renamed the parameters from `start` and `end` to `first` and `last`. The suggestion on style is also adopted.